src/fknni/faiss/faiss.py

```diff
@@ -32,87 +31,156 @@ def __init__(
+        self.features_nan = None
+        self.min_data_ratio = 0.25
+        super().__init__()
 
-    def fit(self, X: np.ndarray | pd.DataFrame, *, y: np.ndarray | None = None) -> FaissImputer:
-        """Fits the FaissImputer to the provided data.
+    @overrides
+    def fit_transform(self, X: np.ndarray, y=None, **fit_params) -> ndarray[Any, dtype[Any]] | None:
+        """Imputes missing values in the data using the fitted Faiss index.
 
         Args:
             X: Input data with potential missing values.
             y: Ignored, present for compatibility with sklearn's TransformerMixin.
 
-        Raises:
-          ValueError: If any parameters are set to an invalid value.
+        Returns:
+            Data with imputed values as a NumPy array of the original data type.
         """
-        X = np.asarray(X, dtype=np.float32)
-        if isinstance(X, pd.DataFrame):
-            X = X.replace(self.missing_values, np.nan).values
-        else:
-            X = np.where(X == self.missing_values, np.nan, X)
-
-        if np.isnan(X).all(axis=0).any():
+        self.X_full = np.asarray(X, dtype=np.float32)
+        if not np.isnan(self.X_full).any():
+            return self.X_full
+        if np.isnan(self.X_full).all(axis=0).any():
             raise ValueError("Features with all values missing cannot be handled.")
 
-        mask = ~np.isnan(X).any(axis=1)
-        X_non_missing = X[mask]
-
-        index = faiss.index_factory(X_non_missing.shape[1], self.index_factory)
-        index.metric_type = faiss.METRIC_L2 if self.metric == "l2" else faiss.METRIC_INNER_PRODUCT
-        index.train(X_non_missing)
-        index.add(X_non_missing)
-
-        self.index_ = index
-        self.global_fallbacks_ = (
-            np.nanmean(X, axis=0) if self.strategy in ["mean", "weighted"] else np.nanmedian(X, axis=0)
+        # Prepare fallback values, used to prefill the query vectors nan´s
+        # or as an imputation fallback if we can't build an index
+        global_fallbacks_ = (
+            np.nanmean(self.X_full, axis=0) if self.strategy in ["mean", "weighted"] else np.nanmedian(self.X_full, axis=0)
         )
 
-        return self
-
-    def transform(self, X: np.ndarray | pd.DataFrame) -> np.ndarray:
-        """Imputes missing values in the data using the fitted Faiss index.
-
-        Args:
-            X: Data with missing values to impute. Expected to be either a NumPy array or a pandas DataFrame.
+        # We will need to impute all features having nan´s
+        feature_indices_to_impute = [i for i in range(self.X_full.shape[1]) if np.isnan(self.X_full[:, i]).any()]
+
+        # Now impute iteratively
+        while feature_indices_to_impute:
+            feature_indices_being_imputed, training_indices, training_data, index = self._fit_train_imputer(feature_indices_to_impute)
+
+            # Use fallback data if we can't build an index and iterate again
+            # TODO: Should we warn the user when this happens?
```

Comment:
I'd say yes although we already have such warnings in ehrapy as well. We'll have to ensure that we don't double them. Oh and let's only warn once.